### PR TITLE
Metrics docs 

### PR DESCRIPTION
**Pull request by williamFalcon**

### CODE REVIEW

File: docs/source/metrics.rst [hunk 1]
```diff
@@ -1,4 +1,296 @@
...
+Metrics are used to monitor model performance.
+
+In this package we provide two major pieces of functionality.
+
+    1. A Metric class you can use to implement metrics with built-in distributed (ddp) support which are device agnostic.
+    2. A collection of popular metrics already implemented for you.
+
+Example:
+
+.. code-block:: python
```
Comment (Borda) [inline docs/source/metrics.rst:13]: lets make this as test

File: docs/source/metrics.rst [hunk 2]
```diff
@@ -1,4 +1,334 @@
...
+Metrics
+=======
+
+Metrics are used to monitor model performance.
+
+In this package we provide two major pieces of functionality.
+
+    1. A Metric class you can use to implement metrics with built-in distributed (ddp) support which are device agnostic.
+    2. A collection of popular metrics already implemented for you.
+
```
Comment (SkafteNicki) [inline docs/source/metrics.rst:10]: Should we mention that the package also comes with a interface to sklearn metrics (with a warning that this is slow due to casting back-and-forth of the tensors)

File: docs/source/metrics.rst [hunk 3]
```diff
@@ -1,4 +1,334 @@
...
+.. warning:: Numpy metrics might slow down your training substantially,
+    since every metric computation requires a GPU sync to convert tensors to numpy.
+
+TensorMetric
+^^^^^^^^^^^^
+Here's an example showing how to implement a TensorMetric
+
+.. code-block:: python
+
+    # @justus TODO: Simple TensorMetric example
```
Comment (SkafteNicki) [inline docs/source/metrics.rst:48]: Simple root mean square error metric:
```suggestion
    class RMSE(TensorMetric):
        def forward(self, x, y):
            return torch.sqrt(torch.mean(torch.pow(x-y, 2.0)))
```

File: docs/source/metrics.rst [hunk 4]
```diff
@@ -1,4 +1,334 @@
...
+.. autoclass:: pytorch_lightning.metrics.metric.TensorMetric
+    :noindex:
+
+NumpyMetric
+^^^^^^^^^^^
+Here's an example showing how to implement a NumpyMetric
+
+.. code-block:: python
+
+    # @justus TODO: Simple NumpyMetric example
```
Comment (SkafteNicki) [inline docs/source/metrics.rst:59]: Same example just in numpy:
```suggestion
    class RMSE(NumpyMetric):
        def forward(self, x, y):
            return np.sqrt(np.mean(np.power(x-y, 2.0)))
        
```

File: docs/source/metrics.rst (hunk 2 above)
Comment (justusschock) [reply docs/source/metrics.rst:10]: We should. We should also include that sklearn has to be installed separately

**Comment by justusschock** [comment]
Maybe we should also show an example on how to parametrize ddp reduction during `__init__`?

And an example where you use a functional (and maybe also with ddp reduction/autocasting via `tensor_metric` function wrapper/decorator?

File: docs/source/metrics.rst (hunk 2 above)
Comment (SkafteNicki) [reply docs/source/metrics.rst:10]: We should make it clear how to import the different backends. Maybe something like:

Use native backend
`import pytorch_lightning.metrics.native as plm`

Use sklearn backend
`import pytorch_lightning.metrics.sklearn as plm`

Use default (native if available else sklearn)
`import pytorch_lightning.metrics as plm`

File: docs/source/metrics.rst [hunk 5]
```diff
@@ -1,4 +1,312 @@
...
+
+.. code-block:: python
+
+    from pytorch_lightning.metrics import Accuracy
+
+    # Plain PyTorch
+    class MyModule(nn.Module):
+        def __init__(self):
+            super().__init__()
+            self.rmse = RMSE()
```
Comment (SkafteNicki) [inline docs/source/metrics.rst:84]: We should probably use accuracy here, since RMSE is not implemented in metrics yet (hope I have time for a PR one of the next days). 
```suggestion
            self.metric = Accuracy()
```

File: docs/source/metrics.rst [hunk 6]
```diff
@@ -1,4 +1,312 @@
...
+
+        def forward(self, x, y):
+            y_hat = # ...
+            acc = self.rmse(y_hat, y)
```
Comment (SkafteNicki) [inline docs/source/metrics.rst:88]:
```suggestion
            acc = self.metric(y_hat, y)
```

File: docs/source/metrics.rst [hunk 7]
```diff
@@ -1,4 +1,312 @@
...
+
+    # PyTorch Lightning
+    class MyModule(pl.LightningModule):
+        def __init__(self):
+            super().__init__()
+            self.rmse = RMSE()
```
Comment (SkafteNicki) [inline docs/source/metrics.rst:94]:
```suggestion
            self.metric = Accuracy()
```

File: docs/source/metrics.rst [hunk 8]
```diff
@@ -1,4 +1,312 @@
...
+
+        def training_step(self, batch, batch_idx):
+            x, y = batch
+            y_hat = # ...
+            acc = self.rmse(y_hat, y)
```
Comment (SkafteNicki) [inline docs/source/metrics.rst:99]:
```suggestion
            acc = self.metric(y_hat, y)
```

File: pytorch_lightning/metrics/classification.py [hunk 9]
```diff
@@ -519,6 +637,17 @@ def __init__(
...
+        Example::
+
+            pred = torch.tensor([0, 1, 2, 3])
+            target = torch.tensor([0, 1, 2, 2])
+            metric = MulticlassROC()
+            classes_roc = metric(pred, target)
+
+        Out::
+
+            # TODO: fix bug - @nicki skafte
```
Comment (williamFalcon) [inline pytorch_lightning/metrics/classification.py:649]: @SkafteNicki

File: pytorch_lightning/metrics/classification.py [hunk 10]
```diff
@@ -569,6 +698,17 @@ def __init__(
...
+        Example::
+
+            pred = torch.tensor([0, 1, 2, 3])
+            target = torch.tensor([0, 1, 2, 2])
+            metric = MulticlassPrecisionRecall()
+            classes_pr = metric(pred, target)
+
+        Out::
+
+            # TODO: fix bug - @nicki skafte
```
Comment (williamFalcon) [inline pytorch_lightning/metrics/classification.py:710]: @SkafteNicki

File: pytorch_lightning/metrics/classification.py [hunk 11]
```diff
@@ -623,6 +763,17 @@ def __init__(
...
+        Example::
+
+            pred = torch.tensor([0, 1, 2, 3])
+            target = torch.tensor([0, 1, 2, 2])
+            metric = DiceCoefficient()
+            classes_pr = metric(pred, target)
+
+        Out::
+
+            # TODO: fix bug - @nicki skafte
```
Comment (williamFalcon) [inline pytorch_lightning/metrics/classification.py:776]: @SkafteNicki

File: docs/source/metrics.rst [hunk 12]
```diff
@@ -1,4 +1,312 @@
...
+handles automated DDP syncing and converts all inputs and outputs to tensors.
+
+.. warning:: Numpy metrics might slow down your training substantially,
+    since every metric computation requires a GPU sync to convert tensors to numpy.
+
+TensorMetric
+^^^^^^^^^^^^
+Here's an example showing how to implement a TensorMetric
+
+.. code-block:: python
```
Comment (Borda) [inline docs/source/metrics.rst:46]:
```suggestion
.. testcode::
```

File: docs/source/metrics.rst [hunk 13]
```diff
@@ -1,4 +1,312 @@
...
+            return torch.sqrt(torch.mean(torch.pow(x-y, 2.0)))
+
+.. autoclass:: pytorch_lightning.metrics.metric.TensorMetric
+    :noindex:
+
+NumpyMetric
+^^^^^^^^^^^
+Here's an example showing how to implement a NumpyMetric
+
+.. code-block:: python
```
Comment (Borda) [inline docs/source/metrics.rst:59]:
```suggestion
.. testcode::
```

File: pytorch_lightning/metrics/functional/classification.py [hunk 14]
```diff
@@ -667,6 +823,28 @@ def dice_score(
         no_fg_score: float = 0.0,
         reduction: str = 'elementwise_mean',
 ) -> torch.Tensor:
+    """
+    @nicki to finish
```
Comment (williamFalcon) [inline pytorch_lightning/metrics/functional/classification.py:827]: @SkafteNicki

Comment (williamFalcon) [reply pytorch_lightning/metrics/functional/classification.py:827]: what are the args to this function?

File: pytorch_lightning/metrics/functional/classification.py [hunk 15]
```diff
@@ -650,6 +791,21 @@ def average_precision(
         sample_weight: Optional[Sequence] = None,
         pos_label: int = 1.,
 ) -> torch.Tensor:
+    """
+
+    Args:
```
Comment (williamFalcon) [inline pytorch_lightning/metrics/functional/classification.py:796]: @SkafteNicki are these args correct?

File: docs/source/metrics.rst (hunk 2 above)
Comment (williamFalcon) [reply docs/source/metrics.rst:10]: @SkafteNicki can you add the sklearn details in here?

File: docs/source/metrics.rst [hunk 16]
```diff
@@ -1,4 +1,315 @@
...
+In this package we provide two major pieces of functionality.
+
+1. A Metric class you can use to implement metrics with built-in distributed (ddp) support which are device agnostic.
+2. A collection of popular metrics already implemented for you.
+
+Example:
+
+.. testcode::
+
+    from pytorch_lightning.metrics.functional import accuracy
```
Comment (williamFalcon) [inline docs/source/metrics.rst:15]: are these imports wrong?
@SkafteNicki @justusschock is there a better way?

File: docs/source/metrics.rst [hunk 17]
```diff
@@ -1,4 +1,315 @@
...
+You can implement metrics as either a PyTorch metric or a Numpy metric. Numpy metrics
+will slow down training, use PyTorch metrics when possible.
+
+Use :class:`TensorMetric` to implement native PyTorch metrics. This class
+handles automated DDP syncing and converts all inputs and outputs to tensors.
+
+Use :class:`NumpyMetric` to implement numpy metrics. This class
+handles automated DDP syncing and converts all inputs and outputs to tensors.
+
+.. warning:: Numpy metrics might slow down your training substantially,
```
Comment (Borda) [inline docs/source/metrics.rst:46]:
```suggestion
.. warning:: 
    Numpy metrics might slow down your training substantially,
```

File: docs/source/metrics.rst [hunk 18]
```diff
@@ -1,4 +1,319 @@
...
+    from pytorch_lightning.metrics import Accuracy
+
+    # Plain PyTorch
+    class MyModule(Module):
+        def __init__(self):
+            super().__init__()
+            self.metric = Accuracy()
+
+        def forward(self, x, y):
+            y_hat = # ...
```
Comment (Borda) [inline docs/source/metrics.rst:94]:
```suggestion
            y_hat = ...
```

File: docs/source/metrics.rst [hunk 19]
```diff
@@ -1,4 +1,319 @@
...
+
+    # PyTorch Lightning
+    class MyModule(LightningModule):
+        def __init__(self):
+            super().__init__()
+            self.metric = Accuracy()
+
+        def training_step(self, batch, batch_idx):
+            x, y = batch
+            y_hat = # ...
```
Comment (Borda) [inline docs/source/metrics.rst:105]:
```suggestion
            y_hat = ...
```

File: pytorch_lightning/metrics/__init__.py [hunk 20]
```diff
@@ -1,30 +1,15 @@
-"""
```
Comment (Borda) [inline pytorch_lightning/metrics/__init__.py:1]: I would keep a short version here...

File: pytorch_lightning/metrics/classification.py [hunk 21]
```diff
@@ -192,11 +224,19 @@ def __init__(
...
+            >>> metric = Precision()
+            >>> metric(pred, target)
+            tensor(0.75)
+
         """
         super().__init__(name='precision',
                          reduce_group=reduce_group,
                          reduce_op=reduce_op)
-
+        # Fixme: bug with result value
```
Comment (Borda) [inline pytorch_lightning/metrics/classification.py:239]: @SkafteNicki mind check this one?

File: docs/source/metrics.rst [hunk 22]
```diff
@@ -1,4 +1,318 @@
...
+
+    pred = torch.tensor([0, 1, 2, 3])
+    target = torch.tensor([0, 1, 2, 2])
+
+    # calculates accuracy across all GPUs and all Nodes used in training
+    accuracy(pred, target)
+
+Out::
+
+    tensor(0.7500)
```
Comment (Borda) [inline docs/source/metrics.rst:29]: this doe s not make sense

File: pytorch_lightning/metrics/classification.py [hunk 23]
```diff
@@ -192,11 +224,18 @@ def __init__(
             reduce_group: the process group to reduce metric results from DDP
             reduce_op: the operation to perform for ddp reduction
 
+        Example:
+
+            >>> pred = torch.tensor([0, 1, 2, 3])
+            >>> target = torch.tensor([0, 1, 2, 2])
+            >>> metric = Precision()
+            >>> metric(pred, target)
+            tensor(1.)
```
Comment (Borda) [inline pytorch_lightning/metrics/classification.py:233]: @williamFalcon is this correct?
